File: app/src/main/java/com/http_s/rest/ProcessingAlert.java
package com.http_s.rest;

import android.app.Activity;
import android.app.AlertDialog;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Handler;
import android.os.SystemClock;
import android.util.DisplayMetrics;
import android.view.View;
import android.view.WindowManager;
import android.webkit.WebView;
import android.widget.TextView;

import org.json.JSONException;
import org.json.JSONObject;


abstract class ProcessingAlert {
    private final Activity activity;
    private final View view;
    private final AlertDialog.Builder alert;
    private AlertDialog OptionDialog;
    private long startTime;
    private final Handler handler;
    private TextView alertTimer;

    private final Runnable runnable = new Runnable() {

        public void run() {
            long millisecondTime = SystemClock.uptimeMillis() - startTime;
            int seconds = (int) (millisecondTime / 1000);
            int minutes = seconds / 60;
            seconds = seconds % 60;
            int milliSeconds = (int) (millisecondTime % 1000);
            if (minutes == 0) {
                alertTimer.setText(seconds + "s, " +
                        String.format("%03d", milliSeconds) + "ms");
            } else {
                alertTimer.setText(minutes + "m, " +
                        seconds + "s, " +
                        String.format("%03d", milliSeconds) + "ms");
            }

            handler.postDelayed(this, 0);
        }

    };

    ProcessingAlert(
            Activity activityInput,
            AlertDialog.Builder alertInput,
            View viewInput) {
        activity = activityInput;
        alert = alertInput;
        view = viewInput;
        handler = new Handler();
        buildLayout();
    }

    private void buildLayout() {
        try {
            alertTimer = view.findViewById(R.id.alertTimer);
            view.findViewById(R.id.cancelRequestButton).setOnClickListener(arg0 -> {
                try {
                    callback(new JSONObject().put("response", "CANCEL_REQUEST"));
                } catch (JSONException e) {
                    e.printStackTrace();
                }
                OptionDialog.dismiss();
            });

            WebView loadingWebView =  view.findViewById(R.id.loadingWebView);
            loadingWebView.getSettings();
            loadingWebView.setBackgroundColor(Color.TRANSPARENT);
            loadingWebView.loadUrl("file:///android_asset/loading.html");

            alert.setView(view);
            OptionDialog = alert.create();
            OptionDialog.setOnDismissListener(dialogInterface -> {
                //TO DO
            });
            //OptionDialog.setCancelable(false);//ALLOW THE DEVICE'S BACK BUTTON TO CLOSE THE ALERT OR NOT

            OptionDialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.parseColor("#00ffffff")));
            OptionDialog.show();

            DisplayMetrics displayMetrics = new DisplayMetrics();
            activity.getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);

            int displayWidth = displayMetrics.widthPixels;
            int displayHeight = displayMetrics.heightPixels;
            WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams();
            layoutParams.copyFrom(OptionDialog.getWindow().getAttributes());

            int dialogWindowWidth = (int) (displayWidth * 0.8f);
            int dialogWindowHeight = (int) (displayHeight * 0.8f);

            layoutParams.width = dialogWindowWidth;
            layoutParams.height = dialogWindowHeight;

            OptionDialog.getWindow().setAttributes(layoutParams);

            startStopWatch();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public AlertDialog.Builder getAlert() {
        return alert;
    }

    public void callback(JSONObject inputResponse) { }

    public void dismiss() {
        OptionDialog.dismiss();
    }

    private void startStopWatch() {
        startTime = SystemClock.uptimeMillis();
        handler.postDelayed(runnable, 0);
    }
}

File: app/src/main/java/com/http_s/rest/alerts/ProcessingAlert.kt
package com.http_s.rest.alerts

import android.app.Activity
import android.app.AlertDialog
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Handler
import android.os.SystemClock
import android.util.DisplayMetrics
import android.view.View
import android.view.WindowManager
import android.webkit.WebView
import android.widget.TextView
import com.http_s.rest.R
import org.json.JSONException
import org.json.JSONObject

internal abstract class ProcessingAlert(
    private val activity: Activity,
    private val alert: AlertDialog.Builder,
    private val view: View
) {
    private lateinit var optionDialog: AlertDialog
    private var startTime: Long = 0
    private val handler = Handler()
    private lateinit var alertTimer: TextView

    private val runnable: Runnable = object : Runnable {
        override fun run() {
            val millisecondTime = SystemClock.uptimeMillis() - startTime
            var seconds = (millisecondTime / 1000).toInt()
            val minutes = seconds / 60
            seconds %= 60
            val milliSeconds = (millisecondTime % 1000).toInt()
            if (minutes == 0) {
                alertTimer.text =
                    seconds.toString() + "s, " + String.format("%03d", milliSeconds) + "ms"
            } else {
                alertTimer.text = minutes.toString() + "m, " +
                        seconds + "s, " + String.format("%03d", milliSeconds) + "ms"
            }

            handler.postDelayed(this, 0)
        }
    }

    init {
        buildLayout()
    }

    private fun buildLayout() {
        try {
            alertTimer = view.findViewById(R.id.alertTimer)
            view.findViewById<View>(R.id.cancelRequestButton).setOnClickListener {
                try {
                    callback(JSONObject().put("response", "CANCEL_REQUEST"))
                } catch (e: JSONException) {
                    e.printStackTrace()
                }
                optionDialog.dismiss()
            }

            val loadingWebView = view.findViewById<WebView>(R.id.loadingWebView)
            loadingWebView.settings
            loadingWebView.setBackgroundColor(Color.TRANSPARENT)
            loadingWebView.loadUrl("file:///android_asset/loading.html")

            alert.setView(view)
            optionDialog = alert.create()
            optionDialog.setOnDismissListener { }

            //OptionDialog.setCancelable(false);//ALLOW THE DEVICE'S BACK BUTTON TO CLOSE THE ALERT OR NOT
            optionDialog.window!!
                .setBackgroundDrawable(ColorDrawable(Color.parseColor("#00ffffff")))
            optionDialog.show()

            val displayMetrics = DisplayMetrics()
            activity.windowManager.defaultDisplay.getMetrics(displayMetrics)

            val displayWidth = displayMetrics.widthPixels
            val displayHeight = displayMetrics.heightPixels
            val layoutParams = WindowManager.LayoutParams()
            layoutParams.copyFrom(optionDialog.window!!.attributes)

            val dialogWindowWidth = (displayWidth * 0.8f).toInt()
            val dialogWindowHeight = (displayHeight * 0.8f).toInt()

            layoutParams.width = dialogWindowWidth
            layoutParams.height = dialogWindowHeight

            optionDialog.window!!.attributes = layoutParams

            startStopWatch()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun getAlert(): AlertDialog.Builder {
        return alert
    }

    open fun callback(inputResponse: JSONObject) {}

    fun dismiss() {
        optionDialog.dismiss()
    }

    private fun startStopWatch() {
        startTime = SystemClock.uptimeMillis()
        handler.postDelayed(runnable, 0)
    }
}
